pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
+
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	InstallIDFileOverride string
+}
+
+// InstallIDData defines the json file that defines installation id.
+type InstallIDData struct {
+	// Time ID was created, in UTC epoch seconds.
+	IDCreatedTimestamp int64 `json:"idCreatedTimestamp"`
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// Stricter than uuid.Parse() which isn't meant for validating strings according
+// to documentation.
+func validInstallID(id string) bool {
+	return regExUUID.MatchString(id)
+}
+
+// SendMetricsSync sends information about application usage. Users can opt out
+// by setting an env variable defined in opt_out.go.
+// Accepts a context for cancellation.
+func SendMetricsSync(ctx context.Context, info *MetricsInfo) error {
+	lookuper := info.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+
+	optOutSettings, err := loadOptOutSettings(ctx, lookuper, info.AppID)
+	if err != nil {
+		return fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	if optOutSettings.NoMetrics {
+		return nil
+	}
+
+	generateNewID := true
+	storedID, err := loadInstallID(info)
+	if err == nil && storedID != nil {
+		oneDayAgo := time.Now().Add(-24 * time.Hour)
+		// Defensively check for case ID Created is in future.
+		generateNewID = oneDayAgo.Unix() >= storedID.IDCreatedTimestamp ||
+			time.Now().Unix() < storedID.IDCreatedTimestamp
+	}
+	var installID string
+	if generateNewID {
+		installUUID, err := uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("could not generate id for metrics: %w", err)
+		}
+		installID = installUUID.String()
+		err = storeInstallID(info, &InstallIDData{
+			IDCreatedTimestamp: time.Now().Unix(),
+			InstallID:          installID,
+		})
+		if err != nil {
+			logging.FromContext(ctx).DebugContext(ctx, "error storing installID", "error", err.Error())
+		}
+	} else {
+		installID = storedID.InstallID
+	}
+
+	var c metricsConfig
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: lookuper,
+	}); err != nil {
+		return fmt.Errorf("failed to process envconfig: %w", err)
+	}
```

Comment:
Didn't we just load the opt-out settings above? I'm confused why we're processing the environment multiple times.

Reply:
Good point. It doesn't really make sense to separate opt-out. I think it was originally written that way because there would be a common opt out shared by metrics and updater, but I think it makes more sense for each of those to have it in their own config.

https://github.com/abcxyz/abc-updater/issues/58

Reply:
(will separate in a future PR, getting alarmingly close to a centithread)